File: java/facade/exercise/OrderFacade.java
package facade.exercise;

public class OrderFacade {
    private InventoryManager inventoryManager;
    private PaymentProcessor paymentProcessor;
    private ShippingLogistics shippingLogistics;

    public OrderFacade() {
        this.inventoryManager = new InventoryManager();
        this.paymentProcessor = new PaymentProcessor();
        this.shippingLogistics = new ShippingLogistics();
    }

    public boolean placeOrder(String productId, int quantity, double amount, String paymentMethod) {
        // Check inventory
        if (!inventoryManager.updateInventory(productId, quantity)) {
            return false;
        }

        // Process payment
        if (!paymentProcessor.processPayment(amount, paymentMethod)) {
            return false;
        }

        // Ship order
        String orderId = generateOrderId(); // Assume orderId generation
        String shippingStatus = shippingLogistics.shipOrder(orderId);
        System.out.println("Shipping Status: " + shippingStatus);

        return true;
    }

    private String generateOrderId() {
        // Logic to generate order id
        return "12345"; // Placeholder value
    }
}

File: PHP/src/Structural/Facade/Exercise/one/OrderFacade.php
<?php

namespace App\Structural\Facade\Exercise\one;

class OrderFacade
{

    private InventoryManagement $inventoryManagement;

    private ShippingLogistics $shippingLogistics;

    private PaymentProcessing $paymentProcessing;

    public function __construct()
    {
        $this->inventoryManagement = new InventoryManagement();
        $this->shippingLogistics = new ShippingLogistics();
        $this->paymentProcessing = new PaymentProcessing();
    }

    /**
     * @param  string $productId
     * @param  int $quantity
     * @param  float $amount
     * @param  string $paymentMethod
     * @return bool
     */
    public function placeOrder(string $productId, int $quantity, float $amount, string $paymentMethod): bool
    {
        if (!$this->inventoryManagement->updateInventory($productId, $quantity)) {
            throw new \Exception("Insufficient inventory");
        }
        if (!$this->paymentProcessing->processPayment($amount, $paymentMethod)) {
            throw new \Exception("Payment failed");
        };
        $orderId = uniqid();
        $this->shippingLogistics->ship($orderId);
        echo "Order " . $orderId . " placed successfully";
        return true;
    }

}
